lib/api/decorators/schema.rb

```diff
@@ -0,0 +1,128 @@
+                                                                    name: title,
+                                                                    required: required,
+                                                                    writable: writable)
+          schema.min_length = min_length if min_length
+          schema.max_length = max_length if max_length
+
+          property property,
+                   getter: -> (*) { schema },
+                   writeable: false
+        end
+
+        def schema_with_allowed_link(property,
+                                          type: make_type(property),
+                                          title: make_title(property),
+                                          href_callback:,
+                                          required: true,
+                                          writable: true)
+          raise ArgumentError if property.nil?
+
+          property property,
+                   exec_context: :decorator,
+                   getter: -> (*) {
+                     representer = ::API::Decorators::AllowedValuesByLinkRepresenter.new(
+                       type: type,
+                       name: title,
+                       required: required,
+                       writable: writable)
+
+                     if represented.defines_assignable_values?
+                       representer.allowed_values_href = instance_eval(&href_callback)
+                     end
+
+                     representer
+                   }
+        end
+
+        def schema_with_allowed_collection(property,
+                                                type: make_type(property),
+                                                title: make_title(property),
+                                                values_callback:,
+                                                value_representer:,
+                                                link_factory:,
+                                                required: true,
+                                                writable: true)
+          raise ArgumentError unless property
+
+          property property,
+                   exec_context: :decorator,
+                   getter: -> (*) {
+                     representer = ::API::Decorators::AllowedValuesByCollectionRepresenter.new(
+                       type: type,
+                       name: title,
+                       current_user: current_user,
+                       value_representer: value_representer,
+                       link_factory: -> (value) { instance_exec(value, &link_factory) },
+                       required: required,
+                       writable: writable)
+
+                     if represented.defines_assignable_values?
+                       representer.allowed_values = instance_exec(&values_callback)
+                     end
+
+                     representer
+                   }
+        end
```

Comment:
An additional note to my short comment above:

> I don't know if it is strictly neccessary to have the `current_user` as instance variable of the representers. However, this is a design decision that was made before I arrived here and I think it would at least take some effort to remove the `current_user` instance method from all representers (which is what we needed to do consequently).

After seeing how stuff works, I am now in favor of removing the `current_user` from the representers. It seems as if it is already causing trouble:

````ruby
# work_package_representer.rb
property :version,
         embedded: true,
         exec_context: :decorator,
         if: ->(*) { represented.fixed_version.present? }

def version
  Versions::VersionRepresenter.new(represented.fixed_version, current_user: current_user)
end
````

Because ROAR is not able to transfer the `current_user` automatically into the contained representer, this has to be done manually.

This is something that now causes headaches for me, too, as I am trying to embed the version of a custom field in a work package.